detection_rules/rule_loader.py

```diff
@@ -4,45 +4,209 @@
 # 2.0.
 
 """Load rule metadata transform between rule and api formats."""
-import functools
 import glob
 import io
 import os
 import re
 from collections import OrderedDict
 from pathlib import Path
-from typing import Dict, List, Iterable
+from typing import Dict, List, Iterable, Callable, Optional
 
 import click
 import pytoml
 
 from .mappings import RtaMappings
 from .rule import RULES_DIR, TOMLRule, TOMLRuleContents, EQLRuleData, KQLRuleData
-from .schemas import CurrentSchema
+from .schemas import CurrentSchema, definitions
 from .utils import get_path, cached
 
 RTA_DIR = get_path("rta")
 FILE_PATTERN = r'^([a-z0-9_])+\.(json|toml)$'
 
 
-def mock_loader(f):
-    """Mock rule loader."""
-    @functools.wraps(f)
-    def wrapped(*args, **kwargs):
+def path_getter(value: str) -> Callable[[dict], bool]:
+    """Get the path from a Python object."""
+    path = value.replace("__", ".").split(".")
+
+    def callback(obj: dict):
+        for p in path:
+            if isinstance(obj, dict) and p in path:
+                obj = obj[p]
+            else:
+                return None
+
+        return obj
+
+    return callback
+
+
+def dict_filter(_obj: Optional[dict] = None, **critieria) -> Callable[[dict], bool]:
+    """Get a callable that will return true if a dictionary matches a set of criteria.
+
+    * each key is a dotted (or __ delimited) path into a dictionary to check
+    * each value is a value or list of values to match
+    """
+    critieria.update(_obj or {})
+    checkers = [(path_getter(k), set(v) if isinstance(v, (list, set, tuple)) else {v}) for k, v in critieria.items()]
+
+    def callback(obj: dict) -> bool:
+        for getter, expected in checkers:
+            target_values = getter(obj)
+            target_values = set(target_values) if isinstance(target_values, (list, set, tuple)) else {target_values}
+
+            return bool(expected.intersection(target_values))
+
+        return False
+
+    return callback
+
+
+def metadata_filter(**metadata) -> Callable[[TOMLRule], bool]:
+    """Get a filter callback based off rule metadata"""
+    flt = dict_filter(metadata)
+
+    def callback(rule: TOMLRule) -> bool:
+        target_dict = rule.contents.metadata.to_dict()
+        return flt(target_dict)
+
+    return callback
+
+
+production_filter = metadata_filter(maturity="production")
+deprecate_filter = metadata_filter(maturity="deprecated")
+
+
+class RuleCollection:
+    """Collection of rule objects."""
+
+    __default = None
+
+    def __init__(self, rules: Optional[List[TOMLRule]] = None):
+        self.id_map: Dict[definitions.UUIDString, TOMLRule] = {}
+        self.file_map: Dict[Path, TOMLRule] = {}
+        self.rules: List[TOMLRule] = []
+        self.frozen = False
+
+        self._toml_cache: Dict[Path, dict] = {}
```

Comment:
to make it a bit more obvious

```suggestion
        self._toml_contents_cache: Dict[Path, dict] = {}
```

(or similar)

Reply:
`_toml_load_cache` ftw?

Reply:
95ad22312c9ded401f6b5f20fb674f9fb413597c